src/legacy/core_plugins/embeddable_api/public/np_ready/public/plugin.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+} from '../../../../../../core/public';
+import { TriggerRegistry, ActionRegistry, EmbeddableFactoryRegistry } from './types';
+import { createEmbeddables, Embeddables } from './api';
+
+export class EmbeddablePublicPlugin implements Plugin<any, any> {
+  private readonly triggers: TriggerRegistry = new Map();
+  private readonly actions: ActionRegistry = new Map();
+  private readonly embeddableFactories: EmbeddableFactoryRegistry = new Map();
+  private embeddables!: Embeddables;
+
+  constructor(initializerContext: PluginInitializerContext) {}
+
+  public setup(core: CoreSetup) {
+    this.embeddables = createEmbeddables({
+      actions: this.actions,
+      embeddableFactories: this.embeddableFactories,
+      triggers: this.triggers,
+    });
+
+    const {
+      registerTrigger,
+      registerAction,
+      registerEmbeddableFactory,
+      attachAction,
+    } = this.embeddables.api;
+
+    return {
+      registerTrigger,
+      registerAction,
+      registerEmbeddableFactory,
+      attachAction,
+    };
+  }
+
+  public start(core: CoreStart) {
+    return this.embeddables.api;
```

Comment:
this seems to allow registration after setup method ? is there any good reason to keep that ? else i would prefer to have that only as part of setup contract.

Reply:
I think it has to be a part of `start` because DashboardContainer registers itself in start because the UICapabilities service is only available in start and dashboard uses it in it's constructor.

Or at least that is why that one function `registerEmbeddableFactory` has to be in start.  Maybe the rest can only be in setup.

Reply:
Discussed 1-on-1 that all Embeddable API needs to be in *start* lifecycle, the only exception&mdash;we probably could remove `registerTrigger` function from the *start* contract; but there is no need of doing that.

For posterity:

- `registerAction` needs to be in *start* lifecycle as actions may depend on `startCore`. Also, in Phase II, Embeddable actions will be dynamic.
- `registerEmbeddableFactory` needs to be in *start* lifecycle as well as there are already use cases that register embeddable factories after the platform has started, because they need to load something.